plugin/externalPluginEntry.go

```diff
@@ -217,121 +185,71 @@ func (e *externalPluginEntry) Stream(ctx context.Context) (io.ReadCloser, error)
 			headerRdrCh <- fmt.Errorf("read an invalid header: %v", string(buf))
 			return
 		}
-
-		// Good to go.
 		headerRdrCh <- nil
 	}()
-
 	timeout := 5 * time.Second
 	timer := time.After(timeout)
-
-	// Waiting for the command to finish ensures that the
-	// stdout + stderr readers are closed
-	//
-	// TODO: Talk about how to handle timeout here, i.e.
-	// whether we should kill the command or not. Should we
-	// do this in a separate goroutine?
-	waitForCommandToFinish := func() {
-		activity.Record(ctx, "Waiting for command: %v", cmdStr)
-		err = cmd.Wait()
-		if err != nil {
-			activity.Record(ctx, "Failed waiting for command: %v", err)
-		}
-	}
-
 	select {
 	case err := <-headerRdrCh:
 		if err != nil {
-			defer waitForCommandToFinish()
-
-			// Try to get more context from stderr, if there is
-			// any
+			defer wait()
+			// Try to get more context from stderr
 			stderr, readErr := ioutil.ReadAll(stderrR)
 			if readErr == nil && len(stderr) != 0 {
 				err = fmt.Errorf(string(stderr))
 			}
-
 			return nil, fmt.Errorf("failed to read the header: %v", err)
 		}
-
-		// Keep reading from stderr so that the streaming isn't
-		// blocked when its buffer is full.
+		// err == nil, meaning we've received the header. Keep reading from
+		// stderr so that the streaming isn't blocked when its buffer is full.
 		go func() {
-			defer func() {
-				activity.Record(ctx, "Closed stream stderr: %v", stderrR.Close())
-			}()
-
-			buf := make([]byte, 4096)
-			for {
-				_, err := stderrR.Read(buf)
-				if err != nil {
-					break
-				}
-			}
+			_, _ = io.Copy(ioutil.Discard, stderrR)
 		}()
-
-		return &stdoutStreamer{stdoutR, cmd}, nil
+		return &stdoutStreamer{cmd, stdoutR}, nil
 	case <-timer:
-		defer waitForCommandToFinish()
-
-		// We timed out while waiting for the streaming header to appear,
-		// so log an appropriate error message using whatever was printed
-		// on stderr
+		defer wait()
+		// We timed out while waiting for the streaming header to appear.
+		// Return an appropriate error message using whatever was printed
+		// on stderr.
 		errMsgFmt := fmt.Sprintf("did not see the %v header after %v seconds:", header, timeout) + " %v"
-
 		stderr, err := ioutil.ReadAll(stderrR)
 		if err != nil {
 			return nil, fmt.Errorf(
 				errMsgFmt,
 				fmt.Sprintf("cannot report reason: stderr i/o error: %v", err),
 			)
 		}
-
 		if len(stderr) == 0 {
 			return nil, fmt.Errorf(
 				errMsgFmt,
 				fmt.Sprintf("cannot report reason: nothing was printed to stderr"),
 			)
 		}
-
 		return nil, fmt.Errorf(errMsgFmt, string(stderr))
 	}
 }
 
-// Exec executes a command on the given entry
 func (e *externalPluginEntry) Exec(ctx context.Context, cmd string, args []string, opts ExecOptions) (ExecCommand, error) {
-	// TODO: Figure out how to pass-in opts when we have entries
-	// besides Stdin. Could do something like
+	// Start the command.
+	//
+	// TODO: Handle other options. Go with
```

Comment:
Does this seem like a good calling convention for Exec's options? Note that `<opts>` will be JSON serialized.

We could also do `<cmd> <args> <opts>` where `<args>` is a JSON array, but that could make things harder for Bash plugin authors (and also less flexible with line-by-line output formats due to the possibility of an arg including a newline).

Reply:
Yeah, that is more idiomatic to how you'd think about an invoked command. It does require some opt-parsing logic though, but we could make that calling convention configurable.